lib/quicly.c

```diff
@@ -3843,11 +3844,20 @@ static int do_send(quicly_conn_t *conn, quicly_send_context_t *s)
         assert(min_packets_to_send <= s->max_datagrams);
 
         if (restrict_sending) {
-            /* PTO (try to send new data when handshake is done, otherwise retire oldest handshake packets and retransmit) */
+            /* PTO: when handshake is in progress, send from the very first unacknowledged byte so as to maximize the chance of
+             * making progress. When handshake is complete, transmit new data if any, else retransmit the oldest unacknowledged data
+             * that is considered inflight. */
             QUICLY_PROBE(PTO, conn, conn->stash.now, conn->egress.sentmap.bytes_in_flight, conn->egress.cc.cwnd,
                          conn->egress.loss.pto_count);
             ++conn->super.stats.num_ptos;
-            if ((ret = mark_packets_on_pto(conn)) != 0)
+            size_t bytes_to_mark = min_packets_to_send * conn->egress.max_udp_payload_size;
+            if (conn->initial != NULL && (ret = mark_frames_on_pto(conn, QUICLY_EPOCH_INITIAL, &bytes_to_mark)) != 0)
+                goto Exit;
+            if (bytes_to_mark != 0 && conn->handshake != NULL &&
+                (ret = mark_frames_on_pto(conn, QUICLY_EPOCH_HANDSHAKE, &bytes_to_mark)) != 0)
+                goto Exit;
+            if (bytes_to_mark != 0 && !scheduler_can_send(conn) &&
```

Comment:
```suggestion
            /**
             * Mark already sent 1-RTT data for PTO only if application does not have new data,
             * i.e., only if scheduler_can_send() returns false.
             */
            if (bytes_to_mark != 0 && !scheduler_can_send(conn) &&
```

Reply:
👍 Thanks for suggestion. Looks good to me.

Reply:
Applied with minor changes in 03bbbf5.